Added support for online synced folder path

pksunkara opened the pull request:
Fixes #351

dirk-thomas commented:
In the context a various different backends to store the actual data being requested I would rather like to see a more scalable approach. The current patch adds various conditions in multiple places. If more backends would be added in the future that seem to likely become unmanageable. Does that make any sense?

---

Since this package is only "sparsely" maintained it is even more important to have a good test coverage to avoid breakage of functionality. Please add test coverage for the newly added feature in order for the PR to be considered for being merged. This can "wait" though until the patch is "feature complete" in order to not waste effort on writing specs when the patch isn't ready to be considered for merging yet.

pksunkara commented:
@dirk-thomas When you say `patch is "feature complete" in order`, you want me to make the backend scalable, right? Or was there anything else? I am going to try to get this ready to be merged.

dirk-thomas commented:
@pksunkara Since a few months ago I have stopped using Atom in favor of a different editor. Hence I won't spend my spare time on this package anymore. Hopefully one of the other users maintaining the package in the past will be able to help. Sorry.

pksunkara commented:
The PR code is now ready to be reviewed:

1. Made a working plugin system
2. Move gist logic to gist plugin
3. Added folder plugin
4. Added specs for the folder plugin

@eimfach Could you review it too?

pksunkara commented:
@dirk-thomas I understand. Thanks for all the work. If no one's interested in maintaining this, I would be willing to do it. Would you mind reviewing this PR, and once this is merged, make me a colloborator?

dirk-thomas commented:
@jerone @groupsky @joefitzgerald @adamrenklint Since all of you have been maintaining the package in the past or have collaboration roles for this repo I wanted to get your opinion too. I don't mind granting access to a potential new maintainer since I won't have the time and interest to fulfill the role anymore.

dirk-thomas commented:
@pksunkara Currently CI is failing for your patch. You might want to look into why that is the case and get it "green" in the meantime.

pksunkara commented:
Fixed the tests. Just a small issue. PR's done and ready

pksunkara commented:
@dirk-thomas Since they are not interested, I think we should move forward.

eimfach commented:
@pksunkara Will have a look at it ! 
@dirk-thomas I am also interested to maintain this repo

jerone commented:
@dirk-thomas I have no opinion against granting others access to this package to keep this package alive. Go for it.

pksunkara commented on lib/plugins/folder.coffee:
```diff
@@ -0,0 +1,78 @@
+  blacklistKeys: (blacklistedKeys) ->
+    blacklistedKeys.push('sync-settings.folderPath')
+
+  initialize: ->
```
It's part of the plugin API. Check the gist plugin.

eimfach replied:
alright

pksunkara commented on lib/plugins/folder.coffee:
```diff
@@ -0,0 +1,78 @@
+    if not @getFolderPath()
+      return notifyMissingMandatorySettings(["Folder Path to sync"])
+
+    console.debug('checking latest backup from synced folder...')
```
They were there from before

eimfach commented:
@pksunkara 
The PR you provide has a lot of changes, also I see that you retouched/rearranged existing code but nothing changed there, which blurs the history of those code lines  (and I don't understand why you did that).
Some explanation from your side e.g. a PR Summary would be very helpful..

It seems a lot to me just for the sake of using a file from fs instead from a url.
